### ISSUE TITLE
NPE when trying to start trace

### ISSUE THREAD

**Issue by dcapwell** [btraceio/btrace#309]
```
$ btrace --version
BTrace v.1.3.10 (20171121)
$ java -version
java version "1.8.0_131"
Java(TM) SE Runtime Environment (build 1.8.0_131-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.131-b11, mixed mode)
$ btrace 53898 TimeMethod.java
! ERROR
java.lang.NullPointerException
        at com.sun.btrace.runtime.Preprocessor.loadEventHandlers(Preprocessor.java:585)
        at com.sun.btrace.runtime.Preprocessor.initRuntime(Preprocessor.java:508)
        at com.sun.btrace.runtime.Preprocessor.processClinit(Preprocessor.java:500)
        at com.sun.btrace.runtime.Preprocessor.process(Preprocessor.java:205)
        at com.sun.btrace.runtime.BTraceProbeNode.initialize(BTraceProbeNode.java:395)
        at com.sun.btrace.runtime.BTraceProbeNode.initialize(BTraceProbeNode.java:377)
        at com.sun.btrace.runtime.BTraceProbeNode.<init>(BTraceProbeNode.java:74)
        at com.sun.btrace.runtime.BTraceProbeFactory.createProbe(BTraceProbeFactory.java:72)
        at com.sun.btrace.agent.Client.load(Client.java:413)
        at com.sun.btrace.agent.Client.loadClass(Client.java:276)
        at com.sun.btrace.agent.Client.loadClass(Client.java:269)
        at com.sun.btrace.agent.RemoteClient.<init>(RemoteClient.java:64)
        at com.sun.btrace.agent.Main.startServer(Main.java:673)
        at com.sun.btrace.agent.Main.access$000(Main.java:60)
        at com.sun.btrace.agent.Main$2.run(Main.java:125)
        at java.lang.Thread.run(Thread.java:748)
```

**Comment by dcapwell** [comment]
cant share the script but here are the probes

```
@OnMethod(clazz="java.lang.System", method="currentTimeMillis")
  public static void onEnter()

@OnMethod(clazz="java.lang.System", method="currentTimeMillis", location = @Location(Kind.RETURN))
  public static void onExit()

@OnEvent
  public static void onEvent()
```

This is my first script.  Trying to get a hello world where I build a histogram of how long methods take.  I picked currentTimeMillis since I know most code calls it so I could profile any random thing running rather than profile something I wrote; this is just for me to learn brace.

**Comment by dcapwell** [comment]
The pid in the example points to `/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk`

```
$ echo $JAVA_HOME
/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home
```

So both processes are using the same JVM

**Comment by dcapwell** [comment]
Based off walking the code, adding the following to unblock me

```
@OnEvent("fluffy")
```

Sending a named event `fluffy` seems to work.

**Comment by dcapwell** [comment]
So I am able to trigger the event and get back most of the body (the code calls println 10 times, output is only 9), but the state is all empty (AtomicLong all valued 0).

So thought maybe some code isn't calling Millies, so took a jstack of the process (IntelliJ) and had the following stack

```
   java.lang.Thread.State: TIMED_WAITING (parking)
        at sun.misc.Unsafe.park(Native Method)
        - parking to wait for  <0x0000000794538740> (a java.util.concurrent.SynchronousQueue$TransferStack)
        at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
        at java.util.concurrent.SynchronousQueue$TransferStack.awaitFulfill(SynchronousQueue.java:460)
        at java.util.concurrent.SynchronousQueue$TransferStack.transfer(SynchronousQueue.java:362)
        at java.util.concurrent.SynchronousQueue.poll(SynchronousQueue.java:941)
        at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1066)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:748)
```

So I know `java.util.concurrent.SynchronousQueue.poll` is getting called so updated to the following probes

```
@OnMethod(clazz="java.util.concurrent.SynchronousQueue", method="poll")
  public static void onEnter()

@OnMethod(clazz="java.util.concurrent.SynchronousQueue", method="poll", location = @Location(Kind.RETURN))
  public static void onExit()
```

Also added a println as the first statement in both probes to verify they get called.  The behavior I see is fluffy still returns 0 for all values, and that the onEnter and onExit methods never print anything.

Should I file a different issue for this?

**Comment by dcapwell** [comment]
Ok more fun.  I dumped the classes out and looking at `javap -v -p -cp /tmp/deleteme/ java.util.concurrent.SynchronousQueue`; there are no references to btrace in the class

**Comment by dcapwell** [comment]
Dumped out the class and see that the wrong method is picked; thing I need to change the method name; will try to find out how

```
public E poll();
    descriptor: ()Ljava/lang/Object;
    flags: ACC_PUBLIC
    Code:
      stack=5, locals=1, args_size=1
         0: invokestatic  #232                // Method $btrace$TimeMethod$2$onEnter:()V
         3: aload_0
         4: getfield      #7                  // Field java/util/concurrent/SynchronousQueue.transferer:Ljava/util/concurrent/SynchronousQueue$Transferer;
         7: aconst_null
         8: iconst_1
         9: lconst_0
        10: invokevirtual #10                 // Method java/util/concurrent/SynchronousQueue$Transferer.transfer:(Ljava/lang/Object;ZJ)Ljava/lang/Object;
        13: invokestatic  #235                // Method java/util/concurrent/SynchronousQueue.$btrace$TimeMethod$2$onExit:()V
        16: areturn
      LineNumberTable:
        line 955: 3
    Signature: #100                         // ()TE;

  public E poll(long, java.util.concurrent.TimeUnit) throws java.lang.InterruptedException;
    descriptor: (JLjava/util/concurrent/TimeUnit;)Ljava/lang/Object;
    flags: ACC_PUBLIC
    Code:
      stack=6, locals=5, args_size=3
         0: invokestatic  #232                // Method $btrace$TimeMethod$2$onEnter:()V
         3: aload_0
         4: getfield      #7                  // Field java/util/concurrent/SynchronousQueue.transferer:Ljava/util/concurrent/SynchronousQueue$Transferer;
         7: aconst_null
         8: iconst_1
         9: aload_3
        10: lload_1
        11: invokevirtual #14                 // Method java/util/concurrent/TimeUnit.toNanos:(J)J
        14: invokevirtual #10                 // Method java/util/concurrent/SynchronousQueue$Transferer.transfer:(Ljava/lang/Object;ZJ)Ljava/lang/Object;
        17: astore        4
        19: aload         4
        21: ifnonnull     30
        24: invokestatic  #11                 // Method java/lang/Thread.interrupted:()Z
        27: ifne          36
        30: aload         4
        32: invokestatic  #235                // Method java/util/concurrent/SynchronousQueue.$btrace$TimeMethod$2$onExit:()V
        35: areturn
        36: new           #12                 // class java/lang/InterruptedException
        39: dup
        40: invokespecial #13                 // Method java/lang/InterruptedException."<init>":()V
        43: athrow
      LineNumberTable:
        line 941: 3
        line 942: 19
        line 943: 30
        line 944: 36
    Exceptions:
      throws java.lang.InterruptedException
    Signature: #103
```

**Comment by dcapwell** [comment]
Ok switching to `@OnMethod(clazz="io.netty.channel.nio.NioEventLoop", method="select", location = @Location(Kind.RETURN))` works.  I still need to figure out how to control the method signature, but I have something working now!

**Comment by dcapwell** [comment]
Ok, so the missing value shows up in the next event; looks like a buffer issue

```
^CPlease enter your option:
        1. exit
        2. send an event
        3. send a named event
        4. flush console output
3
Please enter the event name: fluffy
step: 200
0:  0
1:  0
2:  0
3:  0
4:  0
5:  0
6:  0
7:  0
8:  0
^CPlease enter your option:
        1. exit
        2. send an event
        3. send a named event
        4. flush console output
4

^CPlease enter your option:
        1. exit
        2. send an event
        3. send a named event
        4. flush console output
3
Please enter the event name: fluffy
9:  11
step: 200
0:  0
1:  0
2:  0
3:  0
4:  0
5:  0
6:  0
7:  0
8:  0
```

**Comment by jbachorik** [comment]
The NPE problem is tracked in #305 

For the ability to trace based on method signature you can either use the probe method signature (for capturing the method parameters as well) or just specify `type` argument - eg. `type = "java.lang.String (int, java.util.List)"`

**Comment by dcapwell** [comment]
Thanks for the details.  Will try this out soon!

**Comment by jbachorik** [comment]
Fixed in 1.3.10.2
